java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java

```diff
@@ -0,0 +1,365 @@
+    if (certKeyPair != null) {
+      builder.useTls(certKeyPair.cert, certKeyPair.key);
+    }
+    return builder.build();
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        try (FlightServer flightServer =
+                 getStartServer(location ->
+                     initiateServer(location), 3)) {
+          LOGGER.info("Started " + FlightServer.class.getName() + " as " + flightServer);
+          base.evaluate();
+        } finally {
+          close();
+        }
+      }
+    };
+  }
+
+  private FlightServer getStartServer(CheckedFunction<Location, FlightServer> newServerFromLocation,
+                                      int retries)
+      throws IOException {
+
+    final Deque<ReflectiveOperationException> exceptions = new ArrayDeque<>();
+
+    for (; retries > 0; retries--) {
+      final Location location = Location.forGrpcInsecure(config.getHost(), config.getPort());
+      final FlightServer server = newServerFromLocation.apply(location);
+      try {
+        Method start = server.getClass().getMethod("start");
+        start.setAccessible(true);
+        start.invoke(server);
+        return server;
+      } catch (ReflectiveOperationException e) {
+        exceptions.add(e);
+      }
+    }
+
+    exceptions.forEach(
+        e -> LOGGER.error("Failed to start a new " + FlightServer.class.getName() + ".", e));
+    throw new IOException(exceptions.pop().getCause());
+  }
+
+  /**
+   * Sets a port to be used.
+   *
+   * @return the port value.
+   */
+  public int getPort() {
+    return config.getPort();
+  }
+
+  /**
+   * Sets a host to be used.
+   *
+   * @return the host value.
+   */
+  public String getHost() {
+    return config.getHost();
+  }
+
+  @Override
+  public void close() throws Exception {
+    allocator.getChildAllocators().forEach(BufferAllocator::close);
+    AutoCloseables.close(allocator);
+  }
+
+  /**
+   * Builder for {@link FlightServerTestRule}.
+   */
+  public static final class Builder {
+    private final Properties properties = new Properties();
+    private FlightSqlProducer producer;
+    private Authentication authentication;
+    private CertKeyPair certKeyPair;
+
+    /**
+     * Sets the host for the server rule.
+     *
+     * @param host the host value.
+     * @return the Builder.
+     */
+    public Builder host(final String host) {
+      properties.put(ArrowFlightConnectionConfigImpl.ArrowFlightConnectionProperty.HOST.camelName(),
+          host);
+      return this;
+    }
+
+    /**
+     * Sets a random port to be used by the server rule.
+     *
+     * @return the Builder.
+     */
+    public Builder randomPort() {
+      properties.put(ArrowFlightConnectionConfigImpl.ArrowFlightConnectionProperty.PORT.camelName(),
+          FreePortFinder.findFreeLocalPort());
```

Comment:
FWIW, instead of taking a dependency, just bind to port 0; then you can getPort after starting the server